rllib/algorithms/algorithm.py

```diff
@@ -2560,6 +2758,134 @@ def __setstate__(self, state: dict):
+        policies: Optional[Container[PolicyID]] = None,
+        policy_mapping_fn: Optional[Callable[[AgentID, EpisodeID], PolicyID]] = None,
+        policies_to_train: Optional[
+            Union[
+                Container[PolicyID],
+                Callable[[PolicyID, Optional[SampleBatchType]], bool],
+            ]
+        ] = None,
+    ) -> Tuple[Dict, str]:
+        """Converts a checkpoint directory or object to a proper Algorithm state dict.
+
+        The returned state dict can be used inside self.__setstate__().
+
+        Args:
+            checkpoint: Either the checkpoint directory or AIR Checkpoint object to
+                translate into a proper Algorithm state dict.
+            policies: Optional list/set of PolicyIDs. If not None, only those policies
+                listed here will be included in the returned state. Note that
+                state items such as filters, the `is_policy_to_train` function, as
+                well as the multi-agent `policies` dict will be adjusted as well,
+                based on this arg.
+            policy_mapping_fn: An optional (updated) policy mapping function
+                to include in the returned state.
+            policies_to_train: An optional list of policy IDs to be trained
+                or a callable taking PolicyID and SampleBatchType and
+                returning a bool (trainable or not?).
+
+        Returns:
+             Tuple consisting of a state dict usable within the `self.__setstate__()`
+             method and - if applicable - the "v0" checkpoint file.
+        """
+        # Do we have an old ("v0") single checkpoint file
+        # (`checkpoint_00001/checkpoint-00001`)?
+        v0_checkpoint_file = None
+
+        # `checkpoint` is a str: Could be checkpoint file (older checkpoint versions)
+        # or a directory.
+        if isinstance(checkpoint, str):
+            # Checkpoint is dir: Figure out whether this is an older checkpoint format
+            # (with a `checkpoint-\d+` file in it).
+            if os.path.isdir(checkpoint):
+                for file in os.listdir(checkpoint):
+                    path_file = os.path.join(checkpoint, file)
+                    if os.path.isfile(path_file) and re.match("checkpoint-\\d+", file):
+                        v0_checkpoint_file = path_file
+                        break
+                checkpoint = Checkpoint.from_directory(checkpoint)
+            # Checkpoint is a file: Use as-is (interpreting it as old checkpoint
+            # version).
+            elif os.path.isfile(checkpoint):
+                v0_checkpoint_file = checkpoint
+                checkpoint = Checkpoint.from_directory(os.path.dirname(checkpoint))
+            else:
+                raise ValueError(
+                    f"Given checkpoint ({checkpoint}) not found! Must be a "
+                    "checkpoint directory (or a file for older checkpoint versions)."
+                )
```

Comment:
so essentially, we are guessing the version of the checkpoint give a file/dir path.
can we do this in a util function, so things are more modularized, and can even be tested?
the actually logic should be pretty simple:
1. if dir has a rllib_checkpoint.json file, return the version key.
2. if dir contains a checkpoint-xxx file, return (0,)
3. if path is a file, return (0,)
4. raise.